implementations/rust/ockam/ockam_api/src/kafka/inlet_map.rs

```diff
@@ -0,0 +1,144 @@
+/// with the relative inlet listener socket address.
+/// Also takes care of creating inlets dynamically when they are not present yet.
+#[derive(Debug, Clone)]
+pub(crate) struct KafkaInletMap {
+    inner: Arc<Mutex<KafkaInletMapInner>>,
+}
+
+#[derive(Debug)]
+struct KafkaInletMapInner {
+    broker_map: HashMap<BrokerId, SocketAddr>,
+    port_range: PortRange,
+    current_port: u16,
+    bind_host: String,
+    interceptor_route: Route,
+}
+
+impl KafkaInletMap {
+    pub(crate) fn new(
+        interceptor_route: Route,
+        bind_address: String,
+        port_range: PortRange,
+    ) -> KafkaInletMap {
+        Self {
+            inner: Arc::new(Mutex::new(KafkaInletMapInner {
+                interceptor_route,
+                broker_map: HashMap::new(),
+                current_port: port_range.start(),
+                port_range,
+                bind_host: bind_address,
+            })),
+        }
+    }
+
+    #[cfg(test)]
+    pub(crate) async fn retrieve_inlet(&self, broker_id: BrokerId) -> Option<SocketAddr> {
+        let self_guard = self.inner.lock().await;
+        self_guard.broker_map.get(&broker_id).map(|x| x.clone())
+    }
+
+    ///assert the presence of an inlet for a broker
+    /// on first time it'll create the inlet and return the relative address
+    /// on the second one it'll just return the address
+    pub(crate) async fn assert_inlet_for_broker(
+        &self,
+        context: &mut Context,
+        broker_id: BrokerId,
+    ) -> ockam_core::Result<SocketAddr> {
+        let mut self_guard = self.inner.lock().await;
+        if let Some(address) = self_guard.broker_map.get(&broker_id) {
+            Ok(*address)
+        } else {
+            if self_guard.current_port >= self_guard.port_range.end() {
+                //we don't have any port left for the broker!
+                return Err(Error::new(
+                    Origin::Transport,
+                    Kind::ResourceExhausted,
+                    "reached the upper port range",
+                ));
+            }
+
+            let socket_address = SocketAddr::from_str(&format!(
+                "{}:{}",
+                self_guard.bind_host, self_guard.current_port
+            ))
+            .map_err(|err| Error::new(Origin::Transport, Kind::Invalid, err))?;
+
+            let to = route_to_multiaddr(
+                &self_guard
+                    .interceptor_route
+                    .clone()
+                    .modify()
+                    .append(kafka_outlet_address(broker_id))
+                    .into(),
+            );
+            let to = match to {
+                None => {
+                    return Err(Error::new(
+                        Origin::Transport,
+                        Kind::Invalid,
+                        "cannot convert route to multiaddr",
+                    ));
+                }
+                Some(to) => to,
+            };
+
+            let buffer: Vec<u8> = context
```

Comment:
>  could we replace that with a function call?

That would be ideal and would simplify quite a few things as well as testing, but i don't see any easy way to get a reference to the NodeWorker

Reply:
> Also, converting an ockam route to a multiaddr is technically correct. Still, we should be careful with it, as it may lose some information we expect in that multiaddr later (like `node`, `project`, etc.)

If we could use directly a NodeWorker reference I could avoid the MultiAddress altogether

Reply:
Maybe it can be an abstract dependency? Expressed as a trait maybe? `KafkaInletMap` need a way to spawn inlets, it's fine to provide that functionality as an external dependency

Reply:
I've drafted a prototype:  https://github.com/build-trust/ockam/commit/beac50eb5955d489797744837d13511d5ade165f

I expect to need similar access also for secure channels, I think it's beneficial to be able to access node manager pieces directly, but It works only for services (where we can inject the dependency) and then we are duplicating APIs entry points, which is fine as long as it's clear how they are meant to be used.

I wonder if it's worth having a direct API or if RPCs utilities can do the same. RPCs utilities are scheduled to be moved from the command to the API crate.

Reply:
>  RPCs utilities are scheduled to be moved from the command to the API crate.

Sorry, I double-checked, and there is no issue yet!